File: Recursion/src/main/java/divide_and_conquer/Solution22.java
package divide_and_conquer;

import java.util.*;

public class Solution22 {
    public List<String> generateParenthesis(int n) {

        List<String> ans = new ArrayList<>();

        if (n == 0) {
            ans.add("");
            return ans;
        }

        for (int i = 0; i < n; i++) {

            List<String> firstPart = generateParenthesis(i);
            List<String> secondPart = generateParenthesis(n - i - 1);

            for (String fp : firstPart) {
                for (String sp : secondPart) {
                    String res = "(" + fp + ")" + sp;
                    ans.add(res);
                }
            }
        }
        return ans;

    }

    public static void main(String[] args) {
        Solution22 s = new Solution22();
        List<String> res = s.generateParenthesis(3);

        System.out.println(res);
    }
}

File: Recursion/src/main/python/solution22.py
from typing import List


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        if n == 0:
            return [""]

        ans = []

        for i in range(n):
            first_part = self.generateParenthesis(i)
            second_part = self.generateParenthesis(n - i - 1)

            for fp in first_part:
                for sp in second_part:
                    res = f"({fp}){sp}"
                    ans.append(res)

        return ans


s = Solution()
res = s.generateParenthesis(3)
print(res)
